File: app/src/main/java/com/sonu/app/splash/di/AppComponent.java
package com.sonu.app.splash.di;

import android.app.Application;

import com.sonu.app.splash.MyApp;
import com.sonu.app.splash.bus.AppBus;
import com.sonu.app.splash.bus.AppBusModule;
import com.sonu.app.splash.data.DataManager;
import com.sonu.app.splash.data.DataManagerModule;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dagger.android.AndroidInjector;
import dagger.android.support.AndroidSupportInjectionModule;

/**
 * Created by amanshuraikwar on 18/12/17.
 */

@Singleton
@Component(modules = {AppModule.class,
        DataManagerModule.class,
        AppBusModule.class,
        ActivityBindingModule.class,
        AndroidSupportInjectionModule.class})
public interface AppComponent extends AndroidInjector<MyApp> {

    DataManager getDataManager();
    AppBus getAppBus();

    // Gives us syntactic sugar. we can then do DaggerAppComponent.builder().application(this).build().inject(this);
    // never having to instantiate any modules or say which module we are passing the application to.
    // Application will just be provided into our app graph now.
    @Component.Builder
    interface Builder {

        @BindsInstance
        AppComponent.Builder application(Application application);

        AppComponent build();
    }
}

File: app/src/main/java/io/github/amanshuraikwar/splash/di/AppComponent.kt
package io.github.amanshuraikwar.splash.di

import android.app.Application
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjector
import dagger.android.support.AndroidSupportInjectionModule
import io.github.amanshuraikwar.splash.MyApp
import io.github.amanshuraikwar.splash.bus.AppBus
import io.github.amanshuraikwar.splash.bus.AppBusModule
import io.github.amanshuraikwar.splash.data.DataManager
import io.github.amanshuraikwar.splash.data.DataManagerModule
import io.github.amanshuraikwar.splash.data.DataManagerModuleProvides
import io.github.amanshuraikwar.splash.data.jetpack.JetpackDataManagerModule
import io.github.amanshuraikwar.splash.data.jetpack.JetpackDataManagerProvides
import io.github.amanshuraikwar.splash.data.network.NetworkDataManagerModule
import io.github.amanshuraikwar.splash.data.network.NetworkManagerProvides
import javax.inject.Singleton

/**
 * Created by amanshuraikwar on 07/03/18.
 */
@Singleton
@Component(
        modules = [
            (AppModule::class),
            (NetworkDataManagerModule::class),
            (NetworkManagerProvides::class),
            (JetpackDataManagerModule::class),
            (JetpackDataManagerProvides::class),
            (DataManagerModule::class),
            (DataManagerModuleProvides::class),
            (AppBusModule::class),
            (ActivityBindingModule::class),
            (AndroidSupportInjectionModule::class)])
interface AppComponent : AndroidInjector<MyApp> {

    fun dataManager(): DataManager
    fun appBus(): AppBus

    @Component.Builder interface Builder {

        @BindsInstance fun application(app: Application): Builder
        fun build(): AppComponent
    }
}